src/sql/src/plan/statement/ddl.rs

```diff
@@ -1239,17 +1266,33 @@ pub fn plan_create_source(
             }
         }
         ast::SourceEnvelope::Upsert => {
-            let key_encoding = match encoding.as_ref().and_then(|e| e.key.as_ref()) {
-                None => {
-                    bail_unsupported!(format!("upsert requires a key/value format: {:?}", format))
-                }
-                Some(key_encoding) => key_encoding,
-            };
             // `ENVELOPE UPSERT` implies `INCLUDE KEY`, if it is not explicitly
             // specified.
-            if key_envelope == KeyEnvelope::None {
-                key_envelope = get_unnamed_key_envelope(key_encoding)?;
-            }
+            let key_envelope =
+                // `UPSERT` load generators do not have an encoding so must be special-cased.
+                if let GenericSourceConnection::LoadGenerator(LoadGeneratorSourceConnection {
+                    load_generator: LoadGenerator::Upsert(_),
+                    ..
+                }) = &external_connection
+                {
+                    KeyEnvelope::Named("key".to_string())
```

Comment:
nit - move `"key"` to a constant declared in `storage-types/src/sources/load_generator.rs`